contrib/win32/win32compat/win32_dirent.c

```diff
@@ -36,65 +65,68 @@ DIR * opendir(const char *name)
 		return NULL;
 	}
 
-	// add *.* for Windows _findfirst() search pattern
-	swprintf_s(searchstr, MAX_PATH, L"%s\\*.*", wname);
+	/* add *.* for Windows _findfirst() search pattern */
+	if (swprintf(searchstr, MAX_PATH, L"%s\\*.*", wname) == -1) {
+		/* breached MAX_PATH */
+		errno = ENOTSUP;
+		return NULL;
+	}
 	free(wname);
 
-	if ((hFile = _wfindfirst(searchstr, &c_file)) == -1L) 
+	if ((hFile = _wfindfirst(searchstr, &c_file)) == -1) 
 		return NULL; /* errno is set by _wfindfirst */
-	else {
-		if ((pdir = malloc(sizeof(DIR))) == NULL) {
-			_findclose(hFile);
-			errno = ENOMEM;
-			return NULL;
-		}
-			
-		memset(pdir, 0, sizeof(DIR));
-		pdir->hFile = hFile;
-		memcpy(&pdir->c_file, &c_file, sizeof(c_file));
-		pdir->first = 1;
 
-		return pdir ;
+	if ((pdir = malloc(sizeof(DIR))) == NULL) {
+		_findclose(hFile);
+		errno = ENOMEM;
+		return NULL;
 	}
+			
+	memset(pdir, 0, sizeof(DIR));
+	pdir->hFile = hFile;
+	memcpy(&pdir->c_file, &c_file, sizeof(c_file));
+	pdir->first = 1;
+
+	return pdir ;
 }
 
-/* Close the directory stream DIRP.
-   Return 0 if successful, -1 if not.  */
-int closedir(DIR *dirp)
-{
-    if ( dirp && (dirp->hFile) ) {
-        _findclose( dirp->hFile );
+/* 
+ * Close the directory stream DIRP.
+ * Return 0 if successful, -1 if not.  
+ */
+int 
+closedir(DIR *dirp) {
+    if ( dirp && dirp->hFile) {
+        _findclose(dirp->hFile);
         dirp->hFile = 0;
         free (dirp);
```

Comment:
we have to free(dirp) irrespective of dirp->hFile.. so plz move this out of if loop..